### DISCUSSION TITLE
Flow read data filtering items check if last step is empty

### DISCUSSION THREAD

**Discussion by rickysullivan** [directus/directus#16817] [Archive]
I'm trying to read items from a collection, filtering by id, if there's no result then conditionally trigger the next step:

![image](https://user-images.githubusercontent.com/437480/207585355-08cfc30b-7e30-4f73-ad26-939edcf9f37a.png)

How can I check to see if the array is empty?

**Comment by alexvdvalk** [comment]
Is there a reason you're using a query and not grabbing the item by id directly? eg. /items/addresses/1122850421

If you did this way, you you can follow the "fail" path if that record doesnt exist.

<img width="700" alt="Screenshot 2022-12-14 at 14 09 39" src="https://user-images.githubusercontent.com/21194068/207617133-bb72a915-2f67-4c90-a5f7-2ef6b83fa971.png">

**Reply by rickysullivan** [reply]
![image](https://user-images.githubusercontent.com/437480/207671346-20b9ebc6-163e-402c-a1ab-4a17df7269f9.png)

Results in:

![image](https://user-images.githubusercontent.com/437480/207671726-bf8567b0-1968-4656-ae44-bc4b308edbd1.png)

**Reply by paescuj** [reply]
Seems to work... Do you have any other questions?

**Reply by rickysullivan** [reply]
The 403 is a bit werid. I was expecting perhaps a 404. I was under the impression it wasn't working because of an auth issue.

**Reply by peXed** [reply]
I hope it is ok, if I revive the discussion.

Unforunatly, I ran into the same issue. I cant simply query via IDs, because I don't know the IDs of the items that potentially exists (I get them via a filter).

Is there a way to check if $last is empty or is there even a better way to create some kind of upsert trigger?

**Reply by br41nslug** [reply]
@peXed easiest option i.m.o. is to go with a "run script" operation. doing a simple check like:
```
module.exports = async function(data) {
	if (!data['$last'] || data['$last'].length === 0) {
        throw new Error('No items found');
    }
	return {};
}
```

Giving you the success flow as update and the error flow from this operation as insert.

**Reply by peXed** [reply]
> @peXed easiest option i.m.o. is to go with a "run script" operation. doing a simple check like:
> 
> ```
> module.exports = async function(data) {
> 	if (!data['$last'] || data['$last'].length === 0) {
>         throw new Error('No items found');
>     }
> 	return {};
> }
> ```
> 
> Giving you the success flow as update and the error flow from this operation as insert.

Works perfectly! Thank you.

**Reply by rickysullivan** [reply]
Can someone screenshot what their flow looks like with this script operation in place?

**Reply by paescuj** [reply]
> Can someone screenshot what their flow looks like with this script operation in place?

Sure! What are you looking for specifically?

**Reply by rickysullivan** [reply]
Perhaps a birds eye view. What connects to the script operation and how.

**Reply by paescuj** [reply]
Based on your initial post it could look something like this:
![Screenshot 2023-01-23 at 13-08-48 test](https://user-images.githubusercontent.com/5363448/214036775-6721824d-d270-4dd8-8613-e921e4e604f8.png)
Where the script returns either the data from previous step or throws an error:
```js
module.exports = async function(data) {
  if (!data['$last'] || data['$last'].length === 0) {
    throw new Error('No items found');
  }
  return data['$last'];
}
```
And the query in "Read Data" looking as follows:
```json
{
    "filter": {
        "id": {
            "_eq": "{{$trigger.payload.full_address}}"
        }
    }
}
```

Please note that this just serves as an example! The "Run Script" operator is actually superfluous here since it could also be done like https://github.com/directus/directus/discussions/16817#discussioncomment-4402714 (although as mentioned by you, you might get an 403 with that solution)
